File: src/sys/pkg/lib/omaha-client/src/state_machine/time.rs
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use std::time::{Duration, SystemTime};

/// Convert SystemTime to micro seconds after unix epoch, if the time is before that, return 0.
pub fn time_to_i64(time: SystemTime) -> i64 {
    time.duration_since(SystemTime::UNIX_EPOCH).map(|d| d.as_micros()).unwrap_or(0) as i64
}

/// Convert micro seconds after unix epoch to SystemTime.
pub fn i64_to_time(micros: i64) -> SystemTime {
    SystemTime::UNIX_EPOCH + Duration::from_micros(micros as u64)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_round_trip() {
        assert_eq!(123456789, time_to_i64(i64_to_time(123456789)));
    }
}

File: src/developer/forensics/utils/time.cc
// Copyright 2019 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "src/developer/forensics/utils/time.h"

#include <lib/syslog/cpp/macros.h>
#include <lib/zx/time.h>

#include <ctime>

#include "src/lib/fxl/strings/string_printf.h"

namespace forensics {

std::optional<std::string> FormatDuration(zx::duration duration) {
  if (duration == zx::duration::infinite()) {
    return "inf";
  }

  if (duration < zx::sec(0)) {
    return std::nullopt;
  }

  int64_t d = duration.to_hours() / 24;
  duration -= zx::hour(d * 24);

  int64_t h = duration.to_hours();
  duration -= zx::hour(h);

  int64_t m = duration.to_mins();
  duration -= zx::min(m);

  int64_t s = duration.to_secs();

  return fxl::StringPrintf("%ldd%ldh%ldm%lds", d, h, m, s);
}

timekeeper::time_utc CurrentUtcTimeRaw(timekeeper::Clock* clock) {
  timekeeper::time_utc now_utc;
  // UtcNow returns a non-OK status only if the underlying handle is bad or we don't have sufficient
  // rights to read the clock. CHECK-FAIL if this happens because the C++ runtime is expected to
  // properly set up a clock.
  if (const zx_status_t status = clock->UtcNow(&now_utc); status != ZX_OK) {
    FX_PLOGS(FATAL, status) << "Failed to get current Utc time";
  }

  return now_utc;
}

std::string CurrentUtcTime(timekeeper::Clock* clock) {
  timekeeper::time_utc now_utc = CurrentUtcTimeRaw(clock);
  // std::gmtime expects epoch in seconds.
  const int64_t now_utc_seconds = now_utc.get() / zx::sec(1).get();
  char buffer[32];
  strftime(buffer, sizeof(buffer), "%Y-%m-%d %X %Z", std::gmtime(&now_utc_seconds));
  return std::string(buffer);
}

}  // namespace forensics
